### ISSUE TITLE
Microsoft teams upgrade fails 100% of the time

### ISSUE THREAD

**Issue by rbarbrow** [microsoft/winget-pkgs#18415]
winget upgrade --all

results in an error "Installer failed with exit code: 1638"   if the program is running or not running.

**Comment by jedieaston** [comment]
This has something to do with Teams managing its own updates in a less than kosher way, we've discussed it before in #14155 (and several other places).

The solution to the Teams issues is two-pronged (as I understand it):
- Provide an array of programs that are added to the Add and Remove Programs table for each installer, so WinGet can easily track what should've changed on install/upgrade (microsoft/winget-cli#1073). Since Teams' machine-wide installer puts multiple entries, it should be easier to track this way.
- Disable upgrade for certain applications (like Teams) that manage their own updates so that `winget upgrade --all` will not fail. This should solve your issue. It is currently being tracked in microsoft/winget-cli#1163, so if you give that an upvote hopefully it'll get clocked for v1.1.

Basically, this is a lot of words to say we can't fix this right this second. Sorry about that, hopefully it'll work how you'd expect soon.

**Comment by SignFinder** [comment]
It looks like something strange when winget detecting Microsoft Teams,
![image](https://user-images.githubusercontent.com/299814/123221831-03666a00-d4d8-11eb-9d1c-9df6fd209f4d.png)

**Comment by SignFinder** [comment]
I have found the problems root cause.
There are two different teams client on the machine can be. 
First - located in C:\Program Files (x86)\Teams Installer\ System-wide installer. It is deploying from Teams MSI packet https://docs.microsoft.com/ru-ru/microsoftteams/msi-deployment
The second - which can be installed from from that systemwide or from teams exe installer https://www.microsoft.com/ru-ru/microsoft-teams/download-app - to each user. It is locating in C:\Users\%userprofile%\AppData\Roaming\Microsoft\Windows\Start Menu\Programs
Updating only the second one. But winget detecting first.

**Comment by bshoshany** [comment]
I am experiencing a similar issue as well. I have Teams v1.4.00.11161 installed on my computer, and every time I run `winget upgrade --all` it attempts to upgrade it and fails with exit code 1638. The log file says:

```
Windows Installer reconfigured the product. Product Name: Teams Machine-Wide Installer. Product Version: 1.4.0.11161. Product Language: 1033. Manufacturer: Microsoft Corporation. Reconfiguration success or error status: 1638.

Another version of this product is already installed. Installation of this version cannot continue. To configure or remove the existing version of this product, use Add/Remove Programs on the Control Panel.
```

When I type `winget list teams` there are two versions of Teams in the list:

```
Name            Id              Version      Available    Source
----------------------------------------------------------------
Microsoft Teams Microsoft.Teams 1.4.00.11161
Microsoft Teams Microsoft.Teams 1.4.0.4167   1.4.00.11161 winget
```

However, there is only one version of Teams in Add/Remove Programs, and removing it only removes the latest version, not the previous version. Same with `winget uninstall teams`.

**Comment by chbwien** [comment]
If you get 1638 you help with the `--scope`. Most likely you need to do a
`winget install Microsoft.Teams --scope user`
That worked for me (but doesn't help the `--all` problem)

**Comment by WrinkleRelease** [comment]
If I could uninstall Teams completely, I would, and that would take care of the problem. It keeps re-installing itself though. :(

**Comment by Timmmy** [comment]
Same for Visual c++ redistributables

C:\Users\Tim>winget upgrade Microsoft.VC++2013Redist-x64
Er zijn meerdere geïnstalleerde pakketten gevonden die overeenkomen met de invoer criteria. Verfijn de invoer.
Naam                                                         Id
---------------------------------------------------------------------------------------------------
Microsoft Visual C++ 2013 Redistributable (x64)              Microsoft.VC++2013Redist-x64
Microsoft Visual C++ 2013 Redistributable (x86) - 12.0.40664 {9dff3540-fc85-4ed5-ac84-9e3c7fd8bece}

I'm able to select the second one with --id, but the first is working neither with --id, -e or both

**Comment by vedantmgoyal2009** [comment]
Duplicate of #16155

**Comment by akk-7** [comment]
@jedieaston 
I would like to ask how the progress of the **Microsoft teams upgrade issue** is coming along?

**Comment by jedieaston** [comment]
There is currently a spec PR open for app pinning that should help (https://github.com/microsoft/winget-cli/pull/1894), but implementation hasn't started yet. I believe that's on the roadmap for 1.3? 

Basically, no ETA yet.

**Comment by Trenly** [comment]
Thank you for taking the time to report this issue. In the amount of time the issue has been open, there have been several updates to the WinGet CLI which may have helped mitigate some of these concerns. I encourage you to try out the [latest version](https://github.com/microsoft/winget-cli/releases/latest) of the CLI and see if your issue still persists. 

Additional Notes:
* Applications can now be pinned from upgrade using the `winget pin` command
* Teams now has both the new teams and teams classic, which may help with the upgrade process

###### Close with reason: Stale;

**Comment by saurabh-mhaisekar** [comment]
I am still facing same issue... cannot upgrade `Microsoft.Teams.Classic` 

My current Version of winget cli is `v1.7.11261`
